pyasic/ssh/braiins_os.py

```diff
@@ -93,3 +132,46 @@ async def get_led_status(self):
             str: Status of the LED.
         """
         return await self.send_command("cat /sys/class/leds/'Red LED'/delay_off")
+
+    async def upgrade_firmware(self, file_location: str):
+        """
+        Upgrade the firmware of the BOSMiner device.
+
+        Args:
+            file_location (str): The local file path of the firmware to be uploaded.
+
+        Returns:
+            str: Confirmation message after upgrading the firmware.
+        """
+        try:
+            logger.info("Starting firmware upgrade process.")
+
+            if not file_location:
+                raise ValueError("File location must be provided for firmware upgrade.")
+
+            # Upload the firmware file to the BOSMiner device
+            logger.info(f"Uploading firmware file from {file_location} to the device.")
+            await self.send_command(f"scp {file_location} root@{self.ip}:/tmp/firmware.tar.gz")
+
+            # Extract the firmware file
+            logger.info("Extracting the firmware file on the device.")
+            await self.send_command("tar -xzf /tmp/firmware.tar.gz -C /tmp")
+
+            # Run the firmware upgrade script
+            logger.info("Running the firmware upgrade script on the device.")
+            result = await self.send_command("sh /tmp/upgrade_firmware.sh")
+
+            logger.info("Firmware upgrade process completed successfully.")
+            return result
+        except FileNotFoundError as e:
+            logger.error(f"File not found during the firmware upgrade process: {e}")
+            raise
+        except ValueError as e:
+            logger.error(f"Validation error occurred during the firmware upgrade process: {e}")
+            raise
+        except OSError as e:
+            logger.error(f"OS error occurred during the firmware upgrade process: {e}")
+            raise
+        except Exception as e:
+            logger.error(f"An unexpected error occurred during the firmware upgrade process: {e}", exc_info=True)
+            raise
```

Comment:
This should all be in `pyasic.miners.backends.bos`, not in the SSH handler.

Also, SCP doesn't work on some miners, so you may just have to cat the contents of the file into it.

```
async def upgrade_firmware(self, file: Path):
    with aiofiles.open(file, "r") as f:
        upgrade_contents = await f.read()
    
    await self.ssh.send_command(f"echo {upgrade_contents} > /tmp/firmware.tar && sysupgrade /tmp/firmware.tar")
```

Reply:
Moved this section of code into `pyasic.miners.backends.bos`